### PR TITLE
Add option to create subnet for webdispatcher

### PR DESCRIPTION
**Pull request by KimForss**
### Problem
Some customers want to deploy their web dispatchers and central services to a separate subnet.

### Solution
Create a new optional node in the networking section with the subnet information.

"subnet_app": {
      "is_existing": "false",
      "name": "subnet-app",
      "prefix": "10.11.3.0/24",
      "nsg": {
        "is_existing": "false",
        "name": "nsg-app"
      }
    }        ,
    "subnet_web": {
      "is_existing": "false",
      "name": "subnet-web",
      "prefix": "10.11.4.0/24",
      "nsg": {
        "is_existing": "false",
        "name": "nsg-web"
      }
    }
If that section exists the automation will use it and deploy into that subnet, if it does not exist the web dispatchers will be deployed to the app subnet

### CODE REVIEW

File: deploy/template_samples/application_wo_ha_scssubnet.json [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        },
+        "subnet_web": {
+          "is_existing": "false",
+          "name": "subnet-scs",
+          "prefix": "10.1.4.0/24",
+          "nsg": {
+            "is_existing": "false",
+            "name": "nsg-scs"
+          }
+        }
```
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/template_samples/application_wo_ha_scssubnet.json:69]: Changed it to _scs to be consistent

File: deploy/terraform/modules/app_tier/variables_local.tf [hunk 2]
```diff
@@ -34,6 +34,22 @@ locals {
   sub_app_nsg_arm_id = local.sub_app_nsg_exists ? try(local.var_sub_app_nsg.arm_id, "") : ""
   sub_app_nsg_name   = local.sub_app_nsg_exists ? "" : try(local.var_sub_app_nsg.name, "nsg-app")
 
+  # WEB subnet
+  #If subnet_web is not specified deploy into app subnet
+  var_sub_web_defined   = try(var.infrastructure.vnets.sap.subnet_web,null) == null ? false : true
+  var_sub_web           = try(var.infrastructure.vnets.sap.subnet_web, try(var.infrastructure.vnets.sap.subnet_app, {}))
```
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/terraform/modules/app_tier/variables_local.tf]: True - I just added it as a safety measure - removed it

File: deploy/terraform/modules/app_tier/vm-scs.tf [hunk 3]
```diff
@@ -8,8 +8,8 @@ resource "azurerm_network_interface" "scs" {
 
   ip_configuration {
     name                          = "IPConfig1"
-    subnet_id                     = local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].id : azurerm_subnet.subnet-sap-app[0].id
-    private_ip_address            = cidrhost(local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].address_prefixes[0] : azurerm_subnet.subnet-sap-app[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.scs_vm)
+    subnet_id                     = local.var_sub_web_defined ? (local.sub_web_exists ? data.azurerm_subnet.subnet-sap-web[0].id : azurerm_subnet.subnet-sap-web[0].id) : (local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].id : azurerm_subnet.subnet-sap-app[0].id)
+    private_ip_address            = local.var_sub_web_defined ? cidrhost(local.sub_web_exists ? data.azurerm_subnet.subnet-sap-web[0].address_prefixes[0] : azurerm_subnet.subnet-sap-web[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.scs_vm) : cidrhost(local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].address_prefixes[0] : azurerm_subnet.subnet-sap-app[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.scs_vm)
```
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/terraform/modules/app_tier/vm-scs.tf:12]: I think so - makes life easier if the customer wants to merge subnets at some point

File: deploy/terraform/modules/app_tier/vm-webdisp.tf [hunk 4]
```diff
@@ -8,8 +8,8 @@ resource "azurerm_network_interface" "web" {
 
   ip_configuration {
     name                          = "IPConfig1"
-    subnet_id                     = local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].id : azurerm_subnet.subnet-sap-app[0].id
-    private_ip_address            = cidrhost(local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].address_prefixes[0] : azurerm_subnet.subnet-sap-app[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.web_vm)
+    subnet_id                     = local.var_sub_web_defined ? (local.sub_web_exists ? data.azurerm_subnet.subnet-sap-web[0].id : azurerm_subnet.subnet-sap-web[0].id)  : (local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].id : azurerm_subnet.subnet-sap-app[0].id )
+    private_ip_address            = local.var_sub_web_defined ? cidrhost(local.sub_web_exists ? data.azurerm_subnet.subnet-sap-web[0].address_prefixes[0] : azurerm_subnet.subnet-sap-web[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.web_vm) : cidrhost(local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].address_prefixes[0] : azurerm_subnet.subnet-sap-app[0].address_prefixes[0], tonumber(count.index) + local.ip_offsets.web_vm)
```
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/terraform/modules/app_tier/vm-webdisp.tf:12]: I think so - makes life easier if the customer wants to merge subnets at some point

File: deploy/terraform/modules/app_tier/infrastructure.tf [hunk 5]
```diff
@@ -15,6 +15,23 @@ data "azurerm_subnet" "subnet-sap-app" {
   virtual_network_name = split("/", local.sub_app_arm_id)[8]
 }
 
+# Creates scs subnet of SAP VNET
+resource "azurerm_subnet" "subnet-sap-web" {
+  count                = local.enable_deployment && local.sub_scs_defined ? (local.sub_scs_exists ? 0 : 1) : 0
```
Comment (YunziZhangAzure) [inline deploy/terraform/modules/app_tier/infrastructure.tf:20]: The name and comment are not consistent, which leads to some confusion.
If we do want to provide this functionality, maybe we can use "subnet-sap-scs" as resource name here, which is consistent with local.sub_scs_defined and local.sub_scs_exists.

Comment (KimForss) [reply deploy/terraform/modules/app_tier/infrastructure.tf:20]: I updated the naming to be consistent

Comment (KimForss) [reply deploy/terraform/modules/app_tier/infrastructure.tf:20]: The request of having these in two separate subnets are based on customer requirements. This option allows customers to further segment the network traffic and keep the App Servers more isolated.

File: deploy/terraform/modules/app_tier/vm-scs.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/terraform/modules/app_tier/vm-scs.tf]: This is obsolete, according to Morgan SCS and App should be in the same subnet

File: deploy/template_samples/application_wo_ha_websubnet.json [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        }
+      }
+    }
+  },
+  "jumpboxes": {
+    "windows": [],
+    "linux": []
+  },
+  "application": {
+    "enable_deployment": true,
```
Comment (YunziZhangAzure) [inline deploy/template_samples/application_wo_ha_websubnet.json:78]:
```suggestion
    "enable_deployment": true,
     "sid": "Z01",
```

**Comment by rukawata** [comment]
Close as we will not support scs/ers in different subnet than app servers.

**Pull request closed without merging**